hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +48,272 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+        vec3<Scalar> displacement = pos_const - pos_central;
+
+        // update central particle velocity
+        // note: need to be sure that vector -> quaternion angular momentum conversion is correct
+        vec3<Scalar> linmom_change = mass_const
+                                     * (vec3<Scalar>(h_velocity.data[particle_index])
+                                        - vec3<Scalar>(h_initial_vel.data[idx]));
+        vec3<Scalar> angmom_change_vec = cross(displacement, linmom_change);
+        quat<Scalar> angmom_change = Scalar(2.0) * orientation * quat(0.0, angmom_change_vec);
+
+        // get the current accumulated momentum
+        vec3<Scalar> linmom_accum(h_linmom_accum.data[central_idx]);
+        quat<Scalar> angmom_accum(h_angmom_accum.data[central_idx]);
+        linmom_accum += linmom_change;
+        angmom_accum += angmom_change;
+        // update the alternative arrays
+        h_linmom_accum.data[central_idx]
+            = make_scalar3(linmom_accum.x, linmom_accum.y, linmom_accum.z);
+        h_angmom_accum.data[central_idx] = quat_to_scalar4(angmom_accum);
+        }
+    }
+
+void mpcd::CollisionMethod::finishRigidBodyCollision(uint64_t timestep)
+    {
+    ArrayHandle<Scalar3> h_linmom_accum(m_linmom_accum, access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_angmom_accum(m_angmom_accum, access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::readwrite);
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+
+    // add accumulated momentum to the central particle
+    unsigned int num_total = m_pdata->getN();
+    for (unsigned int idx = 0; idx < num_total; ++idx)
+        {
+        // get accumulated momentum for particle
+        vec3<Scalar> linmom_accum(h_linmom_accum.data[idx]);
+        quat<Scalar> angmom_accum(h_angmom_accum.data[idx]);
+
+        // get velocity, mass, and angular momentum
+        Scalar4 vel_mass = h_velocity.data[idx];
+        quat<Scalar> angmom(h_angmom.data[idx]);
+        vec3<Scalar> vel(vel_mass);
+        const Scalar mass = vel_mass.w;
+
+        // update velocity and angular momentum
+        vec3<Scalar> updated_vel = vel;
+        if (mass > 0)
+            {
+            updated_vel += linmom_accum / mass;
+            }
+        quat<Scalar> updated_angmom = angmom + angmom_accum;
+
+        // save update
+        h_angmom.data[idx] = quat_to_scalar4(updated_angmom);
+        h_velocity.data[idx] = make_scalar4(updated_vel.x, updated_vel.y, updated_vel.z, mass);
+        // clear accumulation
+        h_linmom_accum.data[idx] = make_scalar3(0, 0, 0);
+        h_angmom_accum.data[idx] = make_scalar4(0, 0, 0, 0);
```

Comment:
Don't do this here, instead zero before you accumulate. I'll make a note where that should be done. Doing this also violates the `access_mode::read` contract.
```suggestion
```